### PR TITLE
DHFPROD-7533: Filter for entity types in Graph UI

### PR DESCRIPTION
**Pull request by AkshayBajajML**
Demo: https://drive.google.com/file/d/1SrOX5_9Bq8mwjUD0PAxiKeCbb9oLsJGq/view?usp=sharing

#### Checklist: 
```diff
- Note: do not change the below
```

-  ##### Owner:

- [x] JIRA_ID included in all the commit messages
- [x] PR title is in the format JIRA_ID:Title
- [x] Rebase the branch with upstream
- [x] Squashed all commits into a single commit
- [x] Code passes ESLint tests
- [x] Added Tests
- [x] Run UI tests
  

- ##### Reviewer:

- [x] Reviewed Tests

### CODE REVIEW

**Review by briantang — APPROVED** [review]
Functionality LGTM

**Review by wooldridge — APPROVED** [review]

**Review by wooldridge — CHANGES_REQUESTED** [review]
As discussed, we want the filter behavior to pan the graph such that the filtered node is centered in the view and all the nodes maintain their positions relative to one another. The system should not reposition the filtered node such that it changes position relative to the rest of the nodes.

File: marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-view.tsx [hunk 1]
```diff
@@ -22,16 +23,60 @@ type Props = {
 const GraphView: React.FC<Props> = (props) => {
 
   const {modelingOptions, setSelectedEntity, closeSidePanelInGraphView} = useContext(ModelingContext);
+  const [viewSidePanel, setViewSidePanel] = useState(false);
+  const [nodeSuggested, setNodeSuggested] = useState(["a"]);
+  const [entitySelected, setEntitySelected] = useState("");
+  const [isEntitySelected, setIsEntitySelected] = useState(false);
+
+  useEffect(() => {
```
Comment (xnikhil08) [inline marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-view.tsx:31]: This useEffect was removed in the previous PR for deletion if entity. Any particular reason why we are using this again along with viewSidePanel state variable, which is moved to the context file now.

**Review by xnikhil08 — CHANGES_REQUESTED** [review]

**Comment by xnikhil08** [comment]
@AkshayBajajML - The entities list seems to be not updating in graph view, based on any changes in the table view. PLease check it.

File: marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-view.tsx (hunk 1 above)
Comment (AkshayBajajML) [reply marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-view.tsx]: It is not being used anywhere; I have removed them

**Comment by AkshayBajajML** [comment]
> @AkshayBajajML - The entities list seems to be not updating in graph view, based on any changes in the table view. Please check it.

Fixed it

**Review by xnikhil08 — APPROVED** [review]

**Review by briantang — DISMISSED** [review]

**Review by wooldridge — CHANGES_REQUESTED** [review]
I'm seeing some issues:

1. When I click into the filter box after loading, some nodes rearrange. Is this expected given how things are currently working? (I think it might be due to some hardcoding at the moment.)

https://drive.google.com/file/d/1zIhTHf8D7DpZ57mFSb2KsDiwpT_FGfBT/view

2. When I open the graph view and type "Cust," and then select "NamespacedCustomer," the view is centered on "Customer" (not "NamespacedCustomer") and "NamespacedCustomer" is outlined. I'd expect the view to be centered on "NamespacedCustomer".

https://drive.google.com/file/d/12cjM6YVIZMmceEMBr0rQMdUJ3iaDNhcE/view

3. When I use the filter to select "Customer," "Order," and then "Product," the first entity, "Customer," is centered. But the second and third entities, when filtered, are not. I'd expect every filtered entity to be centered in the graph canvas.

https://drive.google.com/file/d/1UYAhrUF6h0RNMINGBkPFs1Dl0IiF_E91/view

**Comment by AkshayBajajML** [comment]
Fixed the above issues and is ready for reviews.

File: marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-vis/graph-vis.tsx [hunk 2]
```diff
@@ -102,35 +152,75 @@ const GraphVis: React.FC<Props> = (props) => {
...
-                values.borderColor = graphConfig.nodeStyles.selectColor;
-                values.borderWidth = 3;
-              } else if (selected) {
-                values.color = getColor(id);
-                values.borderColor = graphConfig.nodeStyles.selectColor;
-                values.borderWidth = 3;
-              } else if (hovering) {
-                values.color = graphConfig.nodeStyles.hoverColor;
-                values.borderWidth = 0;
+        if (e.entityName === modelingOptions.selectedEntity && props.entitySelected) {
```
Comment (wooldridge) [inline marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-vis/graph-vis.tsx:155]: There is a lot of duplicated code here. I would refactor like this:

const node = {ALL THE PROPERTIES CURRENTLY DUPLICATED}
if (e.entityName === modelingOptions.selectedEntity && props.entitySelected) {
  // ADD SPECIAL BORDER/PHYSICS TO node OBJECT (and anything else that is special)
}

File: marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-view.tsx [hunk 3]
```diff
@@ -22,16 +22,50 @@ type Props = {
...
     height: "20px",
     fill: "currentColor",
   };
 
+  const handleFocus = () => {
+    setNodeSuggested([]);
+  };
+
+  const handleTypeaheadChange = (value: any) => {
+    setEntitySelected(value);
```
Comment (wooldridge) [inline marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-view.tsx:40]: Are these values specific to the filtering feature? If so, can we call them something else to distinguish them from entities selected by a user click? Maybe "entityFiltered" and "isEntityFiltered"?

File: marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-view.tsx [hunk 4]
```diff
@@ -22,16 +22,50 @@ type Props = {
...
+    setIsEntitySelected(false);
+    if (value.length > 2) {
+      Object.keys(props.entityTypes).map((key) => {
+        let obj=nodeSuggested;
+        if (value && !nodeSuggested.includes(props.entityTypes[key]["entityName"]) && props.entityTypes[key]["entityName"].toLowerCase().indexOf(value.toLowerCase()) >= 0) {
+          obj.push(props.entityTypes[key]["entityName"]);
+        }
+        setNodeSuggested(obj);
```
Comment (wooldridge) [inline marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-view.tsx:48]: Is nodeSuggested the items that appear in the filter dropdown menu? If so, can we call them something that describes that, like "filterMenuSuggestions"?

File: marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-view.tsx [hunk 5]
```diff
@@ -22,16 +22,50 @@ type Props = {
...
+      });
+    } else {
+      setNodeSuggested([]);
+    }
+  };
+
+  const sendPayload = (value : any) => {
```
Comment (wooldridge) [inline marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-view.tsx:55]: I think of a "payload" as something that is passed to a backend API. This function is just setting component state, right? How about calling it "handleFilterSelect" instead?

**Review by wooldridge — CHANGES_REQUESTED** [review]
See my inline comments.

File: marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-vis/graph-vis.tsx (hunk 2 above)
Comment (AkshayBajajML) [reply marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-vis/graph-vis.tsx]: Refactored and removed the duplicate code.

File: marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-view.tsx (hunk 4 above)
Comment (AkshayBajajML) [reply marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-view.tsx]: Replaced nodeSuggested  with filterMenuSuggestions

File: marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-view.tsx (hunk 5 above)
Comment (AkshayBajajML) [reply marklogic-data-hub-central/ui/src/components/modeling/graph-view/graph-view.tsx]: Replaced payload with handleFilterSelect

File: marklogic-data-hub-central/ui/e2e/cypress/support/components/model/graph-view-side-panel.tsx [hunk 6]
```diff
@@ -42,6 +42,19 @@ class GraphViewSidePanel {
   getPersonEntityPrefix() {
     return cy.findByTestId("prefix");
   }
+
+  selectGraphViewTab() {
```
Comment (xnikhil08) [inline marklogic-data-hub-central/ui/e2e/cypress/support/components/model/graph-view-side-panel.tsx:46]: This seems like something which should be defined in the parent component Model.tsx within support/pages. Also, looks like there is already a function selectView() within Model.tsx which is being used in few places to select a particular view in Model tile. Let's just use one of these two functions and remove the other one.

**Review by xnikhil08 — CHANGES_REQUESTED** [review]

File: marklogic-data-hub-central/ui/e2e/cypress/support/components/model/graph-view-side-panel.tsx (hunk 6 above)
Comment (AkshayBajajML) [reply marklogic-data-hub-central/ui/e2e/cypress/support/components/model/graph-view-side-panel.tsx]: Thanks for the comment; replaced it with selectView().

**Review by xnikhil08 — APPROVED** [review]

**Review by briantang — APPROVED** [review]

**Pull request merged**